### ISSUE TITLE
not sure what else it wants -- all arguments are provided

### ISSUE THREAD

**Issue by yarikoptic** [poldracklab/mriqc#619]
```shell
$> docker run poldracklab/mriqc --session-id action1 --participant_label sid000021 /home/nastase/social_actions/fmri/1021_actions /home/nastase/social_actions/fmri/1021_actions/derivatives/mriqc participant
/usr/local/miniconda/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
usage: mriqc [-h] [-v]
             [--participant_label [PARTICIPANT_LABEL [PARTICIPANT_LABEL ...]]]
             [-m [{T1w,bold,T2w} [{T1w,bold,T2w} ...]]] [-s SESSION_ID]
             [-r RUN_ID] [--nthreads NTHREADS] [--n_procs N_PROCS]
             [--mem_gb MEM_GB] [--write-graph] [--dry-run]
             [--use-plugin USE_PLUGIN] [--testing] [--hmc-afni] [--hmc-fsl]
             [-f] [--fft-spikes-detector] [-w WORK_DIR]
             [--report-dir REPORT_DIR] [--verbose-reports]
             [--ants-nthreads ANTS_NTHREADS] [--ants-settings ANTS_SETTINGS]
             [--deoblique] [--despike] [--start-idx START_IDX]
             [--stop-idx STOP_IDX] [--correct-slice-timing]
             bids_dir output_dir {participant,group} [{participant,group} ...]
mriqc: error: the following arguments are required: bids_dir, output_dir, analysis_level
```

given that it is `[--participant_label [PARTICIPANT_LABEL [PARTICIPANT_LABEL ...]]]` and thus could have arbitrary number of values, I do not quite see how cmdline args parser  could separate `participant label` from `bids_dir` besides having some option which requires no or strictly 1 value.  In my case (good that I have some) I moved `--session-id` to be before bids_dir for that purpose to get it going (until validator failed and quit because runs have different number of scans... I thought that would be ok... need to read the bible on meaning of the run I guess).  So what is the "right" way to signal a boundary between arguments to e.g. `--participant_label` and `bids_dir`?

**Comment by yarikoptic** [comment]
actually when trying to call it from docker->singularity converted (just singularity pull + cleanup of those .wh files) it still fails to run with the same message:

```shell
$> singularity run  /backup/singularity/mriqc.img --participant_label sid000021 --session-id action1 /home/nastase/social_actions/fmri/1021_actions /home/nastase/social_actions/fmri/1021_actions/derivatives/mriqc participant
usage: mriqc [-h] [--version]
             [--participant_label PARTICIPANT_LABEL [PARTICIPANT_LABEL ...]]
             [--session-id SESSION_ID [SESSION_ID ...]]
             [--run-id RUN_ID [RUN_ID ...]] [--task-id TASK_ID [TASK_ID ...]]
             [-m [{T1w,bold,T2w} [{T1w,bold,T2w} ...]]] [-w WORK_DIR]
             [--report-dir REPORT_DIR] [--verbose-reports] [--write-graph]
             [--dry-run] [--profile] [--use-plugin USE_PLUGIN] [--no-sub]
             [--email EMAIL] [-v] [--webapi-url WEBAPI_URL]
             [--webapi-port WEBAPI_PORT] [--upload-strict] [--n_procs N_PROCS]
             [--mem_gb MEM_GB] [--testing] [-f] [--ica] [--hmc-afni]
             [--hmc-fsl] [--fft-spikes-detector] [--fd_thres FD_THRES]
             [--ants-nthreads ANTS_NTHREADS] [--ants-settings ANTS_SETTINGS]
             [--deoblique] [--despike] [--start-idx START_IDX]
             [--stop-idx STOP_IDX] [--correct-slice-timing]
             bids_dir output_dir {participant,group} [{participant,group} ...]
mriqc: error: the following arguments are required: bids_dir, output_dir, analysis_level

```
```shell
$> /backup/singularity/mriqc.img --version                                                                                                                                                                            
mriqc v0.9.7
```
where is the catch???

**Comment by chrisfilo** [comment]
What if you put optional arguments (--) after the compulsory ones?

On Aug 15, 2017 8:04 AM, "Yaroslav Halchenko" <notifications@github.com>
wrote:

actually when trying to call it from docker->singularity converted (just
singularity pull + cleanup of those .wh files) it still fails to run with
the same message:

$> singularity run  /backup/singularity/mriqc.img --participant_label
sid000021 --session-id action1
/home/nastase/social_actions/fmri/1021_actions
/home/nastase/social_actions/fmri/1021_actions/derivatives/mriqc
participant
usage: mriqc [-h] [--version]
             [--participant_label PARTICIPANT_LABEL [PARTICIPANT_LABEL ...]]
             [--session-id SESSION_ID [SESSION_ID ...]]
             [--run-id RUN_ID [RUN_ID ...]] [--task-id TASK_ID [TASK_ID ...]]
             [-m [{T1w,bold,T2w} [{T1w,bold,T2w} ...]]] [-w WORK_DIR]
             [--report-dir REPORT_DIR] [--verbose-reports] [--write-graph]
             [--dry-run] [--profile] [--use-plugin USE_PLUGIN] [--no-sub]
             [--email EMAIL] [-v] [--webapi-url WEBAPI_URL]
             [--webapi-port WEBAPI_PORT] [--upload-strict] [--n_procs N_PROCS]
             [--mem_gb MEM_GB] [--testing] [-f] [--ica] [--hmc-afni]
             [--hmc-fsl] [--fft-spikes-detector] [--fd_thres FD_THRES]
             [--ants-nthreads ANTS_NTHREADS] [--ants-settings ANTS_SETTINGS]
             [--deoblique] [--despike] [--start-idx START_IDX]
             [--stop-idx STOP_IDX] [--correct-slice-timing]
             bids_dir output_dir {participant,group} [{participant,group} ...]
mriqc: error: the following arguments are required: bids_dir,
output_dir, analysis_level

$> /backup/singularity/mriqc.img --version
mriqc v0.9.7

where is the catch???

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/poldracklab/mriqc/issues/619#issuecomment-322342335>,
or mute the thread

**Comment by chrisfilo** [comment]
BTW different number of scans should be a validator warning not an error.

**Comment by yarikoptic** [comment]
thank you @chrisfilo!

re validator -- do not know which one is error and which one is warning:
```
bids@rolando:/tmp$ cat 1021_actions-validator.log
        1: You have to define 'EchoTime1' and 'EchoTime2' for this file. (code: 15 - ECHO_TIME1-2_NOT_DEFINED)
                /sub-sid000021/ses-raiders/fmap/sub-sid000021_ses-raiders_acq-25mm_phasediff.nii.gz
                        You have to define 'EchoTime1' and 'EchoTime2' for this file. It can be included one of the following locations: /acq-25mm_phasediff.json, /sub-sid000021/sub-sid000021_acq-25mm_phasediff.json, /sub-sid000021/ses-raiders/sub-sid000021_ses-raiders_acq-25mm_phasediff.json, /sub-sid000021/ses-raiders/fmap/sub-sid000021_ses-raiders_acq-25mm_phasediff.json

        1: Not all subjects/sessions/runs have the same scanning parameters. (code: 39 - INCONSISTENT_PARAMETERS)
                /sub-sid000021/ses-raiders/func/sub-sid000021_ses-raiders_task-movie_run-01_bold.nii.gz
                         The most common set of dimensions is: 96,96,48,860 (voxels), This file has the dimensions: 96,96,48,850 (voxels).
                /sub-sid000021/ses-raiders/func/sub-sid000021_ses-raiders_task-movie_run-04_bold.nii.gz
                         The most common set of dimensions is: 96,96,48,860 (voxels), This file has the dimensions: 96,96,48,850 (voxels).
                /sub-sid000120/ses-raiders/func/sub-sid000120_ses-raiders_task-movie_run-01_bold.nii.gz
                         The most common set of dimensions is: 96,96,48,860 (voxels), This file has the dimensions: 96,96,48,850 (voxels).
                /sub-sid000120/ses-raiders/func/sub-sid000120_ses-raiders_task-movie_run-04_bold.nii.gz
                         The most common set of dimensions is: 96,96,48,860 (voxels), This file has the dimensions: 96,96,48,850 (voxels).

        Summary:                   Available Tasks:                        Available Modalities: 
        1413 Files, 21.87GB        actions                                 T1w                   
        2 - Subjects               movie                                   bold                  
        3 - Sessions               TODO: full task name for actions        T2w                   
                                   TODO: full task name for movie          fieldmap 
```
but I just also was mislead -- output of mriqc about "BIDS-compliance" has nothing to do with validator probably --just that it couldn't figure things out (e.g. as when I didn't specify that there is a session).
I still have no luck making singularity behave and the box which has docker is already busy with that 'successfull' run and can't even rerun to test placement of '--' args after mandatory ones. 

But with singularity one 
- indeed placing --options after seems to help (Somewhat)!  my rotten soul was not thinking out side of the --help/docs box ;)
- but then can't make sense from behavior reporting that there is no T1w images -- parsing of cmdline args seems to be ok:

```
$> singularity run -B /home/nastase:/mnt /backup/singularity/mriqc-ed.img /mnt/social_actions/fmri/1021_actions /mnt/social_actions/fmri/1021_actions/derivatives/mriqc participant --session-id actions1 --participant_label sid000021
> /usr/local/miniconda/lib/python3.6/site-packages/mriqc/bin/mriqc_run.py(164)main()
-> opts = get_parser().parse_args()
(Pdb) n
> /usr/local/miniconda/lib/python3.6/site-packages/mriqc/bin/mriqc_run.py(167)main()
-> log_level = int(max(3 - opts.verbose_count, 0) * 10)
(Pdb) p opts
Namespace(analysis_level=['participant'], ants_nthreads=1, ants_settings=None, bids_dir='/mnt/social_actions/fmri/1021_actions', correct_slice_timing=False, deoblique=False, despike=False, dry_run=False, email='', fd_thres=0.2, fft_spikes_detector=False, float32=False, hmc_afni=True, hmc_fsl=False, ica=False, mem_gb=0, modalities=['T1w', 'bold', 'T2w'], n_procs=0, no_sub=False, output_dir='/mnt/social_actions/fmri/1021_actions/derivatives/mriqc', participant_label=['sid000021'], profile=False, report_dir=None, run_id=None, session_id=['actions1'], start_idx=None, stop_idx=None, task_id=None, testing=False, upload_strict=False, use_plugin=None, verbose_count=0, verbose_reports=False, webapi_port=None, webapi_url='https://mriqc.nimh.nih.gov/api/v1', work_dir='/backup/singularity/work', write_graph=False)
(Pdb) import glob
(Pdb) glob.glob('/mnt/social_actions/fmri/1021_actions/sub*/ses*/anat/*')
['/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-actions1/anat/sub-sid000120_ses-actions1_acq-MPRAGE_T1w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-actions1/anat/sub-sid000120_ses-actions1_acq-MPRAGE_T1w.json', '/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-raiders/anat/sub-sid000120_ses-raiders_T2w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-raiders/anat/sub-sid000120_ses-raiders_T2w.json', '/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-actions2/anat/sub-sid000120_ses-actions2_acq-MPRAGE_T1w.json', '/mnt/social_actions/fmri/1021_actions/sub-sid000120/ses-actions2/anat/sub-sid000120_ses-actions2_acq-MPRAGE_T1w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-actions1/anat/sub-sid000021_ses-actions1_acq-MPRAGE_T1w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-actions1/anat/sub-sid000021_ses-actions1_acq-MPRAGE_T1w.json', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-actions2/anat/sub-sid000021_ses-actions2_acq-MPRAGE_T1w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-actions2/anat/sub-sid000021_ses-actions2_acq-MPRAGE_T1w.json', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-raiders/anat/sub-sid000021_ses-raiders_T2w.nii.gz', '/mnt/social_actions/fmri/1021_actions/sub-sid000021/ses-raiders/anat/sub-sid000021_ses-raiders_T2w.json']
(Pdb) c
2017-08-15 01:28:21,148 mriqc.cli:WARNING No T1w scans were found in /mnt/social_actions/fmri/1021_actions
2017-08-15 01:28:21,149 mriqc.cli:WARNING No bold scans were found in /mnt/social_actions/fmri/1021_actions
2017-08-15 01:28:21,149 mriqc.cli:WARNING No T2w scans were found in /mnt/social_actions/fmri/1021_actions
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/mriqc", line 11, in <module>
    load_entry_point('mriqc==0.9.7', 'console_scripts', 'mriqc')()
  File "/usr/local/miniconda/lib/python3.6/site-packages/mriqc/bin/mriqc_run.py", line 167, in main
    log_level = int(max(3 - opts.verbose_count, 0) * 10)
RuntimeError: None of the supplied labels (--participant_label) matched with the participants found in the BIDS directory (/mnt/social_actions/fmri/1021_actions).
```

**Comment by yarikoptic** [comment]
I think I see the gotcha there (not sure how worked in docker -- might be having different versions? yeap -- docker one is 0.9.1, freshly pulled singularity 0.9.7):
```shell
 35  	    if session:
 36  ->	        for mod in modalities:
 37  	            queries[mod]['session'] = [session]
 38  	
 39  	    if run:
 40  	        for mod in modalities:
 41  	            queries[mod]['run'] = run
(Pdb) p session
['actions1']
(Pdb) n
> /usr/local/miniconda/lib/python3.6/site-packages/mriqc/utils/bids.py(37)collect_bids_data()
-> queries[mod]['session'] = [session]
(Pdb) n
> /usr/local/miniconda/lib/python3.6/site-packages/mriqc/utils/bids.py(36)collect_bids_data()
-> for mod in modalities:
(Pdb) p queries[mod]
{'modality': 'anat', 'type': 'T1w', 'extensions': ['nii', 'nii.gz'], 'session': [['actions1']]}
```
so the list of sessions gotten from cmdline and help suggests that multiple could be provided 
```
  --session-id SESSION_ID [SESSION_ID ...]
                        select a specific session to be processed
```
so code assumptions of it being a single one seems to be wrong... I guess should look through the subject/sessions pairs?

**Comment by yarikoptic** [comment]
ah -- I see that there were post release fixes already e.g. merged here eed631e1d4dd9e62bcc2c981be62d8c688334f3e  so the next question would I guess be if a "release" coming? ;)

**Comment by oesteban** [comment]
Hi @yarikoptic, answer is yes... but I'm very behind schedule with this one - https://github.com/poldracklab/mriqc/milestone/18
